File: tags/milestone1/Noxa.Emulation.Psp.Bios.NativeHLE/KIntHandler.cpp
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

#include "Stdafx.h"
#include "Kernel.h"
#include "KernelHelpers.h"
#include "KIntHandler.h"
#include <malloc.h>
#include <string.h>

using namespace System;
using namespace Noxa::Emulation::Psp;
using namespace Noxa::Emulation::Psp::Bios;

KIntHandler::KIntHandler( Bios::Kernel* kernel, int interruptNumber, int slot, uint address, uint argument )
{
	Kernel = kernel;
	InterruptNumber = interruptNumber;
	Slot = slot;
	Address = address;
	Argument = argument;
}

KIntHandler::~KIntHandler()
{
	if( _isInstalled == true )
		this->SetEnabled( false );
}

void KIntHandler::SetEnabled( bool enabled )
{
	if( _isInstalled == enabled )
		return;

	assert( Kernel->Cpu != NULL );
	if( _isInstalled == false )
	{
		Kernel->Cpu->RegisterInterruptHandler( InterruptNumber, Slot, Address, Argument );
		_isInstalled = true;
	}
	else
	{
		Kernel->Cpu->UnregisterInterruptHandler( InterruptNumber, Slot );
		_isInstalled = false;
	}
}

File: trunk/Noxa.Emulation.Psp.Bios.ManagedHLE/Kernel Types/KIntHandler.cs
// ----------------------------------------------------------------------------
// PSP Player Emulation Suite
// Copyright (C) 2006 Ben Vanik (noxa)
// Licensed under the LGPL - see License.txt in the project root for details
// ----------------------------------------------------------------------------

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

namespace Noxa.Emulation.Psp.Bios.ManagedHLE
{
	class KIntHandler : KHandle, IDisposable
	{
		public Kernel Kernel;

		public int InterruptNumber;
		public int Slot;

		public uint Address;
		public uint Argument;

		public uint CallCount;

		// Timing info?
		// total clock (int64), min clock (int64), max clock (int64)?

		private bool _isInstalled;

		public KIntHandler( Kernel kernel, int interruptNumber, int slot, uint address, uint argument )
		{
			Kernel = kernel;

			InterruptNumber = interruptNumber;
			Slot = slot;

			Address = address;
			Argument = argument;

			CallCount = 0;
		}

		~KIntHandler()
		{
			this.Dispose();
		}

		public void Dispose()
		{
			GC.SuppressFinalize( this );

			this.Enabled = false;
		}

		public bool Enabled
		{
			get
			{
				return _isInstalled;
			}
			set
			{
				if( value == _isInstalled )
					return;

				if( value == true )
				{
					// Install
					Kernel.Cpu.RegisterInterruptHandler( InterruptNumber, Slot, Address, Argument );
				}
				else
				{
					// Uninstall
					Kernel.Cpu.UnregisterInterruptHandler( InterruptNumber, Slot );
				}
				_isInstalled = value;
			}
		}
	}
}
